subsys/bluetooth/controller/ll_sw/ull_adv_aux.c

```diff
@@ -104,3 +512,259 @@ u8_t ll_adv_aux_set_clear(void)
+	LL_ASSERT(!ret);
+}
+
+static int init_reset(void)
+{
+	/* Initialize adv aux pool. */
+	mem_init(ll_adv_aux_pool, sizeof(struct ll_adv_aux_set),
+		 sizeof(ll_adv_aux_pool) / sizeof(struct ll_adv_aux_set),
+		 &adv_aux_free);
+
+	return 0;
+}
+
+static inline struct ll_adv_aux_set *aux_acquire(void)
+{
+	return mem_acquire(&adv_aux_free);
+}
+
+static inline void aux_release(struct ll_adv_aux_set *aux)
+{
+	mem_release(aux, &adv_aux_free);
+}
+
+static inline u16_t aux_handle_get(struct ll_adv_aux_set *aux)
+{
+	return mem_index_get(aux, ll_adv_aux_pool,
+			     sizeof(struct ll_adv_aux_set));
+}
+
+static void mfy_aux_offset_get(void *param)
+{
+	struct ll_adv_set *adv = param;
+	struct ll_adv_aux_set *aux;
+	u32_t ticks_to_expire;
+	u32_t ticks_current;
+	struct pdu_adv *pdu;
+	u8_t ticker_id;
+	u8_t retry;
+	u8_t id;
+
+	aux = (void *)HDR_LLL2EVT(adv->lll.aux);
+	ticker_id = TICKER_ID_ADV_AUX_BASE + aux_handle_get(aux);
+
+	id = TICKER_NULL;
+	ticks_to_expire = 0U;
+	ticks_current = 0U;
+	retry = 4U;
+	do {
+		u32_t volatile ret_cb = TICKER_STATUS_BUSY;
+		u32_t ticks_previous;
+		u32_t ret;
+
+		ticks_previous = ticks_current;
+
+		ret = ticker_next_slot_get(TICKER_INSTANCE_ID_CTLR,
+					   TICKER_USER_ID_ULL_LOW,
+					   &id,
+					   &ticks_current, &ticks_to_expire,
+					   ticker_op_cb, (void *)&ret_cb);
+		if (ret == TICKER_STATUS_BUSY) {
+			while (ret_cb == TICKER_STATUS_BUSY) {
+				ticker_job_sched(TICKER_INSTANCE_ID_CTLR,
+						 TICKER_USER_ID_ULL_LOW);
+			}
+		}
+
+		LL_ASSERT(ret_cb == TICKER_STATUS_SUCCESS);
+
+		LL_ASSERT((ticks_current == ticks_previous) || retry--);
+
+		LL_ASSERT(id != TICKER_NULL);
+	} while (id != ticker_id);
+
+	/* NOTE: as remainder not used in scheduling primary PDU
+	 * packet timer starts transmission after 1 tick hence the +1.
+	 */
+	aux->lll.ticks_offset = ticks_to_expire + 1;
+
+	pdu = lll_adv_data_curr_get(&adv->lll);
+	lll_adv_aux_offset_fill(ticks_to_expire, 0, pdu);
+}
+
+static void ticker_cb(u32_t ticks_at_expire, u32_t remainder, u16_t lazy,
+		      void *param)
+{
+	static memq_link_t link;
+	static struct mayfly mfy = {0, 0, &link, NULL, lll_adv_aux_prepare};
+	static struct lll_prepare_param p;
+	struct ll_adv_aux_set *aux = param;
+	struct lll_adv_aux *lll;
+	u32_t ret;
+	u8_t ref;
+
+	DEBUG_RADIO_PREPARE_A(1);
+
+	lll = &aux->lll;
+
+	/* Increment prepare reference count */
+	ref = ull_ref_inc(&aux->ull);
+	LL_ASSERT(ref);
+
+	/* Append timing parameters */
+	p.ticks_at_expire = ticks_at_expire;
+	p.remainder = remainder;
+	p.lazy = lazy;
+	p.param = lll;
+	mfy.param = &p;
+
+	/* Kick LLL prepare */
+	ret = mayfly_enqueue(TICKER_USER_ID_ULL_HIGH,
+			     TICKER_USER_ID_LLL, 0, &mfy);
+	LL_ASSERT(!ret);
+
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	struct ll_adv_set *adv;
+
+	adv = (void *)HDR_LLL2EVT(lll->adv);
+	if (adv->lll.sync) {
+		struct ll_adv_sync_set *sync;
+
+		sync  = (void *)HDR_LLL2EVT(adv->lll.sync);
+		if (sync->is_started) {
+			ull_adv_sync_offset_get(adv);
```

Comment:
Doesn't calling this function after the prepare-mayfly has already been enqueued cause a race condition?

Reply:
Not in nRF, as the PDU is picked at start/anchor. But yes, if the PDU is not populated in time, then its an issue. I think this be addressed in separate GH issue/PR to synchronize PDU population during prepare timespace.

Reply:
Create a GH issue for future

Reply:
Create a GH issue for future